### PR TITLE
Add user defined constitutive function

### PR DESCRIPTION
**Pull request by mschreter**
This PR is a follow-up of #2, introducing an interface function for a user-defined material law.

### CODE REVIEW

**Comment by peterrum** [comment]
@mschreter Could you rebase this?

**Comment by mschreter** [comment]
@peterrum done!

File: include/adaflo/navier_stokes.h [hunk 1]
```diff
@@ -168,6 +168,13 @@ class NavierStokes : public FlowBaseAlgorithm<dim>
                            const unsigned int                                face,
                            const double                                      density);
 
+  void
+  set_user_defined_material(std::function<Tensor<2, dim, VectorizedArray<double>>(
+                              const Tensor<2, dim, VectorizedArray<double>> &,
+                              const unsigned int,
+                              const unsigned int,
+                              const bool)> my_user_defined_material);
```
Comment (peterrum) [inline include/adaflo/navier_stokes.h]: Could you document somewhere the meaning of the individual parameters?

File: source/navier_stokes_matrix.cc [hunk 2]
```diff
@@ -848,45 +847,60 @@ NavierStokesMatrix<dim>::local_operation(
...
-                  FlowParameters::newtonian_compressible_stokes_hypothesis)
-                grad_u[d][d] -= tmu_times_2 * divergence / static_cast<double>(dim);
-
-              if (LocalOps == NavierStokesOps::vmult ||
-                  LocalOps == NavierStokesOps::residual)
-                grad_u[d][d] -= pres;
+              Tensor<2, dim, vector_t> grad_u_temp =
+                convert_to_tensor<2, dim, vector_t>(velocity.get_gradient(q));
+              grad_u =
+                tau1 * user_defined_material(grad_u_temp, cell, q, q == 0 /*do_reinit*/);
```
Comment (peterrum) [inline source/navier_stokes_matrix.cc]: How is this supposed to be used? `do_reinit` indicates that the user might be using `FEvaluation` in the function? In that case you have to know fe_index/quad_index there. Is that known there?

File: source/navier_stokes_matrix.cc [hunk 3]
```diff
@@ -848,45 +847,60 @@ NavierStokesMatrix<dim>::local_operation(
...
             }
+          else
+            AssertThrow(false, ExcMessage("Requested ConstitutiveType not found."));
+
+          // add pressure
+          if (LocalOps == NavierStokesOps::vmult || LocalOps == NavierStokesOps::residual)
+            for (unsigned int d = 0; d < dim; ++d)
+              grad_u[d][d] -= pres;
```
Comment (peterrum) [inline source/navier_stokes_matrix.cc:907]: I am wondering why the formatting here is different!?

File: include/adaflo/navier_stokes.h (hunk 1 above)
Comment (mschreter) [reply include/adaflo/navier_stokes.h]: Of course :facepalm:

File: source/navier_stokes_matrix.cc (hunk 2 above)
Comment (mschreter) [reply source/navier_stokes_matrix.cc]: I'll provide a clearer description and prepare an exemplary usage of the interface in the evening.

File: include/adaflo/navier_stokes.h [hunk 4]
```diff
@@ -168,6 +168,19 @@ class NavierStokes : public FlowBaseAlgorithm<dim>
                            const unsigned int                                face,
                            const double                                      density);
 
+  // Set a user defined material law to compute the viscous stress constribution. You have
+  // to pass a lambda function, which computes @return the shear stress rate depending on the
+  // @p velocity_gradient for a given cell @p cell_idx at the quadrature point @p quad_idx.
+  // @p do_tangent=true means that the function is called from the tangent (vmult);
+  // do_tangent=false means that the residual is computed.
```
Comment (mschreter) [inline include/adaflo/navier_stokes.h:175]: @peterrum I replaced the `do_reinit` flag by `do_tangent`, to get the information whether the residual or vmult is computed. I hope the description makes the interface clear now.

**Review by peterrum — APPROVED** [review]

**Pull request merged**